components/brave_rewards/browser/rewards_service_impl.cc

```diff
@@ -275,6 +275,13 @@ std::string GetPrefPath(const std::string& name) {
   return base::StringPrintf("%s.%s", pref_prefix, name.c_str());
 }
 
+template <typename Callback, typename... Args>
+void DeferCallback(base::Location location, Callback callback, Args&&... args) {
+  base::SequencedTaskRunnerHandle::Get()->PostTask(
+      location,
+      base::BindOnce(std::move(callback), std::forward<Args>(args)...));
+}
+
```

Comment:
Just wondering if we might want to define this in a utility file somewhere, as it seems generally useful.

Reply:
Maybe? In general, I dislike utility files because either you end up with a grab-bag of random things or you have a dozen utility files with one thing each. It's not rewards-specific, so if we find a use for it elsewhere we'll have to think of a "shared" place to put it.

Reply:
I'm not sure I understand the reentrancy problems this solves?
Also, I think that instead of deferring the execution of these callbacks with default arguments explicitly, we should create them in a way they know how to execute themselves in case they don't get called. Pretty much how the [mojo::WrapCallbackWithDefaultInvokeIfNotRun<>()](https://source.chromium.org/chromium/chromium/src/+/main:mojo/public/cpp/bindings/callback_helpers.h;l=113-120) wrapper works.

Reply:
In general, callers of an asynchronous API will assume that a callback will not be called synchronously (i.e. control will not be transferred to the callback before the original call has completed). In practice, this will mostly work out OK, but it's better to avoid the surprise.

Is your suggestion that all callers be updated to wrap their callbacks "if not run"? I wouldn't want that to be the caller's responsibility. Or are you suggesting that the method itself wrap the callback as the first thing it does?

I suppose that since `RewardsService` could get destroyed, the callbacks could always not get run, regardless, and the "if not run" wrapper would solve that.

Reply:
Why will callers of an asynchronous API assume anything regarding when their callback will be called? Am I missing some Chromium-only threading-/task-related knowledge here? 🙂

As for `WrapCallbackWithDefaultInvokeIfNotRun<>()`, that's a tough one. I can see why you don't want this to be the callers' responsibility, at the same time, I can't think of a better place. It's also tricky to provide reasonable defaults on a general basis.

Reply:
Programmers will tend to assume the "expected" control flow path, where the callback is not executed synchronously:

```cpp
state_ = 1;

DoSomethingAsync("hello", base::BindOnce([](base::WeakPtr<Self> self) {
  if (self) {
    self->state_ = 2;
  }
}, weak_factory_.GetWeakPtr());

LOG(INFO) << "What is state? " << state_;
```

which can result to hard-to-spot bugs.

I'm interested in the `WrapCallbackWithDefaultInvokeIfNotRun` suggestion though - I'll look into it.

Reply:
Hmm, I mean, that's the programmer's fault here, in my opinion, but I agree that we should avoid these if we can. 👍🏻